src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -151,6 +168,249 @@ void ReadFromRemote::enforceAggregationInOrder()
     DB::enforceAggregationInOrder(stage, *context);
 }
 
+ASTSelectQuery & getSelectQuery(ASTPtr ast)
+{
+    if (const auto * explain = ast->as<ASTExplainQuery>())
+        ast = explain->getExplainedQuery();
+
+    return ast->as<ASTSelectQuery &>();
+}
+
+/// This is an attempt to convert filters (pushed down from the plan optimizations) from ActionsDAG back to AST.
+/// It should not be needed after we send a full plan for distributed queries.
+static ASTPtr tryBuildAdditionalFilterAST(
+    const ActionsDAG & dag,
+    const std::unordered_set<std::string> & projection_names,
+    const std::unordered_map<std::string, QueryTreeNodePtr> & execution_name_to_projection_query_tree,
+    Tables * external_tables,
+    const ContextPtr & context)
+{
+    std::unordered_map<const ActionsDAG::Node *, ASTPtr> node_to_ast;
+
+    struct Frame
+    {
+        const ActionsDAG::Node * node;
+        size_t next_child = 0;
+    };
+    std::stack<Frame> stack;
+    stack.push({dag.getOutputs().front()});
+    while (!stack.empty())
+    {
+        auto & frame = stack.top();
+        const auto * node = frame.node;
+
+        if (node_to_ast.contains(node))
+        {
+            stack.pop();
+            continue;
+        }
+
+        /// Labmdas are not supported (converting back to AST is complicated).
+        /// We have two cases here cause function with no capture can be constant-folded.
+        if (WhichDataType(node->result_type).isFunction()
+            || (node->type == ActionsDAG::ActionType::FUNCTION
+                && typeid_cast<const FunctionCapture *>(node->function_base.get())))
+        {
+            node_to_ast[node] = nullptr;
+            continue;
+        }
+
+        /// Support for IN. The stored AST from the Set is taken.
+        if (WhichDataType(node->result_type).isSet())
+        {
+            auto maybe_set = node->column;
+            if (const auto * col_const = typeid_cast<const ColumnConst *>(maybe_set.get()))
+                maybe_set = col_const->getDataColumnPtr();
+
+            if (const auto * col_set = typeid_cast<const ColumnSet *>(maybe_set.get()))
+                node_to_ast[node] = col_set->getData()->getSourceAST();
```

Comment:
We substitute subquery here, not sending the set to the shard node?

Reply:
Yes